File: src/statesman/commands/SceneCommand.java
package statesman.commands;

import statesman.Content;
import statesman.Interpreter;

public class SceneCommand extends Command {

    public static final String ID_SCENE = "scene";

    private String _targetScene;
    private boolean _hasKey;

    private SceneCommand() {
        _targetScene = "";
        _hasKey = false;
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new SceneCommand();
        }
        return _defaultInstance;
    }

    public SceneCommand(String targetScene) {
        if (targetScene.isEmpty()) {
            throw new IllegalArgumentException();
        }
        _targetScene = targetScene;
        _hasKey = false;
    }

    @Override
    public Command fromText(String commandId, String[] arguments) {
        if (arguments.length == 2) {
            return new SceneCommand(arguments[1]);
        }
        return null;
    }

    @Override
    public void execute() {
        if (_hasKey
                || Content.getScript().getScenes().containsKey(_targetScene)) {
            Interpreter.setScene(
                    Content.getScript().getScenes().get(_targetScene));
            _hasKey = true;
        }
    }

}

File: src/Commands/SceneCommand.cs
namespace Statesman.Commands
{
    public class SceneCommand : Command
    {
        public const string kIdScene = "scene";

        private readonly string _targetScene;
        private bool _hasKey;

        public SceneCommand(string targetScene)
        {
            if (string.IsNullOrWhiteSpace(targetScene))
            {
                throw new ArgumentException("Target scene name cannot be empty");
            }
            _targetScene = targetScene;
            _hasKey = false;
        }

        public new static Command FromText(string commandId, string[] arguments)
        {
            if (commandId != kIdScene)
            {
                return null;
            }
            if (arguments.Length == 2)
            {
                return new SceneCommand(arguments[1]);
            }
            return null;
        }

        public override void Execute()
        {
            if (_hasKey || Content.Script.Scenes.ContainsKey(_targetScene))
            {
                Interpreter.Scene = Content.Script.Scenes[_targetScene];
                _hasKey = true;
            }
        }
    }
}
